core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -680,19 +680,23 @@ class UnifiedLog(@volatile var logStartOffset: Long,
   }
 
   private[log] def activeProducersWithLastSequence: Map[Long, Int] = lock synchronized {
```

Comment:
Is it right that this is only used in tests? If so, can we move this to `LogTestUtils`? Then we don't have to worry about the `toMap` copy.

Reply:
Right, it is used only in tests. But we should not move to `LogTestUtils` as this method should take a lock while building the results. It is better to keep these semantics and leave this method here even though it is used only in tests. This will keep the method contract clean and thread safe.

Reply:
Makes sense about keeping the implementation in `UnifiedLog`. I still think we should simply return `collection.Map` and avoid the copy.